api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto

```diff
@@ -0,0 +1,105 @@
+// [#extension: envoy.filters.thrift.payload_to_metadata]
+
+message PayloadToMetadata {
+  enum ValueType {
+    STRING = 0;
+    NUMBER = 1;
+  }
+
+  // [#next-free-field: 6]
+  message KeyValuePair {
+    // The namespace — if this is empty, the filter's namespace will be used.
+    string metadata_namespace = 1;
+
+    // The key to use within the namespace.
+    string key = 2 [(validate.rules).string = {min_len: 1}];
+
+    oneof value_type {
+      // The value to pair with the given key.
+      //
+      // When used for on_present case, if value is non-empty it'll be used instead
+      // of the field value. If both are empty, the field value is used as-is.
+      //
+      // When used for on_missing case, a non-empty value must be provided.
+      string value = 3;
+
+      // If present, the header's value will be matched and substituted with this.
+      // If there is no match or substitution, the field value is used as-is.
+      //
+      // This is only used for on_present.
+      type.matcher.v3.RegexMatchAndSubstitute regex_value_rewrite = 4;
+    }
+
+    // The value's type — defaults to string.
+    ValueType type = 5 [(validate.rules).enum = {defined_only: true}];
+  }
+
+  // A Rule defines what metadata to apply when a field is present or missing.
+  // [#next-free-field: 6]
+  message Rule {
+    oneof match_specifier {
+      option (validate.required) = true;
+
+      // If specified, the route must exactly match the request method name. As a special case,
+      // an empty string matches any request method name.
+      string method_name = 1;
+
+      // If specified, the route must have the service name as the request method name prefix.
+      // As a special case, an empty string matches any service name. Only relevant when service
+      // multiplexing.
+      string service_name = 2;
+    }
+
+    // Specifies that a match will be performed on the value of a field.
+    FieldSelector field_selector = 3 [(validate.rules).message = {required: true}];
+
+    // If the field is present, apply this metadata KeyValuePair.
+    KeyValuePair on_present = 4;
+
+    // If the field is missing, apply this metadata KeyValuePair.
+    //
+    // The value in the KeyValuePair must be set, since it'll be used in lieu
+    // of the missing field value.
+    KeyValuePair on_missing = 5;
+  }
+
+  message FieldSelector {
+    // field name to log
+    string name = 1 [(validate.rules).string = {min_len: 1}];
+
+    // field id to match. Limit to the range of int16 to fit thrift format.
+    google.protobuf.Int32Value id = 2 [
+      (validate.rules).int32 = {lte: 32767 gte: -32768},
+      (validate.rules).message = {required: true}
+    ];
+
+    // next node of the field selector
+    FieldSelector child = 3;
+  }
+
+  // The list of rules to apply to requests.
+  repeated Rule request_rules = 1 [(validate.rules).repeated = {min_items: 1}];
```

Comment:
is there any value in reusing some of these definitions from the existing header-to-metadata filter?

Reply:
Great question, I'm happy this is raised.
The most reusable part is `message KeyValuePair` and I did think through if we want to DIY or not.

I'm not sure how we handled api dependency between filters. Tag @mattklein123 since he's assigned as the api shepherd.

It's kinda weird to let p2m depends on h2m, since we need to build both filters if we only want to use p2m.
However extract `message KeyValuePair` to something like `common/key_value.proto` would break the existing usage. 
IMO the `KeyValuePair` per se is not that big, so it might be okay-ish to dup that in two filters.
If we really want to DIY in proto, we might want to keep both `KeyValuePair` in `h2m.proto` and  ``common/key_value.proto`` and mark `KeyValuePair` in h2m as deprecated, maintain two code paths to make both work and remove one path in some day. It apparently needs a great amount of care. 

That's the reason I dup `KeyValuePair` in p2m and h2m. Does it make sense?